File: app/src/main/java/cc/lixiaoyu/wanandroid/ui/activity/LoginActivity.java
package cc.lixiaoyu.wanandroid.ui.activity;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import cc.lixiaoyu.wanandroid.R;

import cc.lixiaoyu.wanandroid.base.BaseSwipeBackActivity;
import cc.lixiaoyu.wanandroid.ui.fragment.LoginFragment;
import cc.lixiaoyu.wanandroid.ui.fragment.RegisterFragment;

public class LoginActivity extends BaseSwipeBackActivity {

    private FragmentManager mFragmentManager;
    private FragmentTransaction mTransaction;

    private Fragment loginFragment;
    private Fragment registerFragment;
    private Fragment currentFragment;
    @Override
    protected void initData() {
        loginFragment = LoginFragment.newInstance();
        registerFragment = RegisterFragment.newInstance();
    }

    @Override
    protected void initView() {
        mFragmentManager = getSupportFragmentManager();
        mTransaction = mFragmentManager.beginTransaction();
        mTransaction.add(R.id.login_container, loginFragment, LoginFragment.LOGIN_FRAGMENT_TAG)
                    .add(R.id.login_container, registerFragment, RegisterFragment.REGISTER_FRAGMENT_TAG)
                    .hide(registerFragment)
                    .commit();
    }

    @Override
    protected int attachLayout() {
        return R.layout.activity_login;
    }

    /**
     * 在Activity中控制Fragment的切换
     * @param fromTag   当前fragment的tag
     * @param toTag     目标fragment的tag
     */
    public void switchFragment(String fromTag, String toTag){
        Fragment from = mFragmentManager.findFragmentByTag(fromTag);
        Fragment to  = mFragmentManager.findFragmentByTag(toTag);
        if(currentFragment != to){
            currentFragment = to;
            FragmentTransaction transaction = mFragmentManager.beginTransaction();
            if(!to.isAdded()){
                transaction.hide(from).add(R.id.login_container, to, toTag).commit();
            }else{
                transaction.hide(from).show(to).commit();
            }
        }
    }
}

File: app/src/main/java/cc/lixiaoyu/wanandroid/core/account/ui/LoginActivity.kt
package cc.lixiaoyu.wanandroid.core.account.ui

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import cc.lixiaoyu.wanandroid.R

class LoginActivity : AppCompatActivity() {

    private var signInFragment: Fragment? = null
    private var signUpFragment: Fragment? = null

    private lateinit var viewModel: LoginViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        viewModel = ViewModelProvider(this)[LoginViewModel::class.java]
        initFragment(savedInstanceState)
    }

    private fun initFragment(savedInstanceState: Bundle?) {
        savedInstanceState?.let {
            signInFragment = supportFragmentManager.findFragmentByTag(SIGN_IN_FRAGMENT_TAG)
            signUpFragment = supportFragmentManager.findFragmentByTag(SIGN_UP_FRAGMENT_TAG)
        }
        if (signInFragment == null) {
            signInFragment = SignInFragment.newInstance()
        }
        if (signUpFragment == null) {
            signUpFragment = SignUpFragment.newInstance()
        }
        val transaction = supportFragmentManager.beginTransaction()
        if (!signInFragment!!.isAdded) {
            transaction.add(R.id.login_container, signInFragment!!, SIGN_IN_FRAGMENT_TAG)
        }
        if (!signUpFragment!!.isAdded) {
            transaction.add(R.id.login_container, signUpFragment!!, SIGN_UP_FRAGMENT_TAG)
        }
        val currentFragment = if (viewModel.currentFragmentTag == SIGN_IN_FRAGMENT_TAG) signInFragment!! else signUpFragment!!
        val hiddenFragment = if (viewModel.currentFragmentTag == SIGN_IN_FRAGMENT_TAG) signUpFragment!! else signInFragment!!
        transaction
            .hide(hiddenFragment)
            .show(currentFragment)
            .commitAllowingStateLoss()
    }

    /**
     * 在Activity中控制Fragment的切换
     * @param toTag     目标fragment的tag
     */
    fun switchFragment(toTag: String) {
        if (toTag == viewModel.currentFragmentTag) {
            return
        }
        val toFragment = supportFragmentManager.findFragmentByTag(toTag) ?: return
        val currentFragment = if (viewModel.currentFragmentTag == SIGN_IN_FRAGMENT_TAG) signInFragment else signUpFragment ?: return
        supportFragmentManager.beginTransaction()
            .hide(currentFragment!!)
            .show(toFragment)
            .commitAllowingStateLoss()
        viewModel.currentFragmentTag = toTag
    }

    companion object {
        const val SIGN_IN_FRAGMENT_TAG = "sign_in_fragment"
        const val SIGN_UP_FRAGMENT_TAG = "sign_up_fragment"
    }
}